controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -86,15 +113,88 @@
         this.counter = new AtomicInt96();
         this.limit = new AtomicInt96();
         this.refreshFutureRef = null;
+        this.completedTxnGC = new ZKGarbageCollector(COMPLETED_TXN_GC_NAME, storeHelper, this::gcCompletedTxn, gcPeriod);
+        this.completedTxnGC.startAsync();
+        this.completedTxnGC.awaitRunning();
     }
 
     private void initialize() {
         METRICS_PROVIDER.start();
     }
 
+    private CompletableFuture<Void> gcCompletedTxn() {
+        return storeHelper.getChildren(COMPLETED_TX_BATCH_ROOT_PATH)
+                .thenApply(children -> {
+                            // retain latest two and delete remainder.
+                            List<Long> list = children.stream().map(Long::parseLong).sorted().collect(Collectors.toList());
+                            if (list.size() > 2) {
+                                return list.subList(0, list.size() - 2);
+                            } else {
+                                return new ArrayList<Long>();
+                            }
+                        }
+                )
+                // region Backward Compatibility.
+                // TODO 2755 retire code in this region 
+                .thenCompose(toDeleteList -> {
+                    CompletableFuture<Void> future;
+                    if (!Config.DISABLE_COMPLETED_TXN_BACKWARD_COMPATIBILITY) {
```

Comment:
I think we need to think more carefully about backwards compatibility.
Here and above we have totally different code paths with if/else branches. It would be better if there were an abstraction that we plugged into and instantiated an appropriate class based on for example a stored version.